src/modules/csm/checkpoint.py

```diff
@@ -50,48 +55,46 @@ def __init__(
             raise ValueError("Left border epoch should be less or equal right border epoch")
         self.converter = converter
         self.l_epoch = l_epoch
-        self.r_epoch = self._get_adjusted_r_epoch(r_epoch, finalized_epoch)
+        self.r_epoch = r_epoch
 
-    def __iter__(self):
+        self.max_available_epoch_to_check = min(
+            self.r_epoch, EpochNumber(finalized_epoch - self.CHECKPOINT_SLOT_DELAY_EPOCHS)
+        )
 
+        if self.r_epoch > self.max_available_epoch_to_check and not self._is_min_step_reached(finalized_epoch):
+            raise MinStepIsNotReached()
+
+    def __iter__(self):
         duty_epochs = cast(list[EpochNumber], list(sequence(self.l_epoch, self.r_epoch)))
 
         checkpoint_epochs = []
         for index, epoch in enumerate(duty_epochs, 1):
             checkpoint_epochs.append(epoch)
-            if index % self.MAX_CHECKPOINT_STEP == 0 or epoch == self.r_epoch:
-                # We need to get the first slot of the next of next epoch to get fit to
-                # 8192 roots in `checkpoint_slot` state block_roots to check duties in every epoch in checkpoint.
-                # To check duties in the current epoch you need to
-                # get 32 slots of the current epoch and 32 slots of the next epoch.
-                checkpoint_slot = SlotNumber(self.converter.get_epoch_last_slot(EpochNumber(epoch + 1)) + 1)
+            if epoch == self.max_available_epoch_to_check or index % self.MAX_CHECKPOINT_STEP == 0:
+                checkpoint_slot = self.converter.get_epoch_first_slot(
+                    EpochNumber(epoch + self.CHECKPOINT_SLOT_DELAY_EPOCHS)
+                )
                 logger.info(
                     {"msg": f"Checkpoint slot {checkpoint_slot} with {len(checkpoint_epochs)} duty epochs is prepared"}
                 )
                 yield Checkpoint(checkpoint_slot, checkpoint_epochs)
                 checkpoint_epochs = []
+            if epoch == self.max_available_epoch_to_check:
+                break
```

Comment:
Since we have python 3.12 we can significantly simplify this function.

```python
    def __iter__(self):
        for checkpoint_epochs in batched(
            sequence(self.l_epoch, self.max_available_epoch_to_check),
            self.MAX_CHECKPOINT_STEP,
        ):
            checkpoint_slot = self.converter.get_epoch_first_slot(
                EpochNumber(max(checkpoint_epochs) + self.CHECKPOINT_SLOT_DELAY_EPOCHS)
            )
            logger.info(
                {"msg": f"Checkpoint slot {checkpoint_slot} with {len(checkpoint_epochs)} duty epochs is prepared"}
            )
            yield Checkpoint(checkpoint_slot, checkpoint_epochs)
```